File: backend/src/models/lead_generation.py
"""Lead generation workflow models (US-939).

Pydantic models for ICP definition, discovered leads,
score breakdowns, pipeline views, and outreach.
"""

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class ReviewStatus(str, Enum):
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"
    SAVED = "saved"


class PipelineStage(str, Enum):
    PROSPECT = "prospect"
    QUALIFIED = "qualified"
    OPPORTUNITY = "opportunity"
    CUSTOMER = "customer"


# ICP Models
class ICPDefinition(BaseModel):
    industry: list[str] = Field(default_factory=list)
    company_size: dict[str, int] = Field(default_factory=lambda: {"min": 0, "max": 0})
    modalities: list[str] = Field(default_factory=list)
    therapeutic_areas: list[str] = Field(default_factory=list)
    geographies: list[str] = Field(default_factory=list)
    signals: list[str] = Field(default_factory=list)
    exclusions: list[str] = Field(default_factory=list)


class ICPResponse(BaseModel):
    id: str
    user_id: str
    icp_data: ICPDefinition
    version: int
    created_at: datetime
    updated_at: datetime


# Score Breakdown
class ScoreFactor(BaseModel):
    name: str
    score: int = Field(ge=0, le=100)
    weight: float = Field(ge=0, le=1)
    explanation: str


class LeadScoreBreakdown(BaseModel):
    overall_score: int = Field(ge=0, le=100)
    factors: list[ScoreFactor]


# Discovered Lead Models
class DiscoveredLeadResponse(BaseModel):
    id: str
    user_id: str
    icp_id: str | None
    company_name: str
    company_data: dict[str, object]
    contacts: list[dict[str, object]]
    fit_score: int
    score_breakdown: LeadScoreBreakdown | None
    signals: list[str]
    review_status: ReviewStatus
    reviewed_at: datetime | None
    source: str
    lead_memory_id: str | None
    created_at: datetime
    updated_at: datetime


class LeadReviewRequest(BaseModel):
    action: ReviewStatus = Field(..., description="Review action: approved, rejected, or saved")


# Pipeline Models
class PipelineStageSummary(BaseModel):
    stage: PipelineStage
    count: int
    total_value: float


class PipelineSummary(BaseModel):
    stages: list[PipelineStageSummary]
    total_leads: int
    total_pipeline_value: float


# Outreach Models
class OutreachRequest(BaseModel):
    subject: str = Field(..., min_length=1, max_length=200)
    message: str = Field(..., min_length=1)
    tone: str = Field(default="professional")


class OutreachResponse(BaseModel):
    id: str
    lead_id: str
    draft_subject: str
    draft_body: str
    status: str
    created_at: datetime


# Discovery trigger
class DiscoverLeadsRequest(BaseModel):
    target_count: int = Field(default=10, ge=1, le=50)

File: frontend/src/api/leadGeneration.ts
import { apiClient } from "./client";

// Types
export interface ICPDefinition {
  industry: string[];
  company_size: { min: number; max: number };
  modalities: string[];
  therapeutic_areas: string[];
  geographies: string[];
  signals: string[];
  exclusions: string[];
}

export interface ICPResponse {
  id: string;
  user_id: string;
  icp_data: ICPDefinition;
  version: number;
  created_at: string;
  updated_at: string;
}

export type ReviewStatus = "pending" | "approved" | "rejected" | "saved";
export type PipelineStage = "prospect" | "qualified" | "opportunity" | "customer";

export interface ScoreFactor {
  name: string;
  score: number;
  weight: number;
  explanation: string;
}

export interface LeadScoreBreakdown {
  overall_score: number;
  factors: ScoreFactor[];
}

export interface DiscoveredLead {
  id: string;
  user_id: string;
  icp_id: string | null;
  company_name: string;
  company_data: Record<string, unknown>;
  contacts: Record<string, unknown>[];
  fit_score: number;
  score_breakdown: LeadScoreBreakdown | null;
  signals: string[];
  review_status: ReviewStatus;
  reviewed_at: string | null;
  source: string;
  lead_memory_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface PipelineStageSummary {
  stage: PipelineStage;
  count: number;
  total_value: number;
}

export interface PipelineSummary {
  stages: PipelineStageSummary[];
  total_leads: number;
  total_pipeline_value: number;
}

export interface OutreachRequest {
  subject: string;
  message: string;
  tone?: string;
}

export interface OutreachResponse {
  id: string;
  lead_id: string;
  draft_subject: string;
  draft_body: string;
  status: string;
  created_at: string;
}

// API functions
export async function saveICP(icp: ICPDefinition): Promise<ICPResponse> {
  const response = await apiClient.post<ICPResponse>("/leads/icp", icp);
  return response.data;
}

export async function getICP(): Promise<ICPResponse | null> {
  const response = await apiClient.get<ICPResponse | null>("/leads/icp");
  return response.data;
}

export async function discoverLeads(
  targetCount: number = 10
): Promise<DiscoveredLead[]> {
  const response = await apiClient.post<DiscoveredLead[]>(
    "/leads/discovered",
    { target_count: targetCount }
  );
  return response.data;
}

export async function listDiscoveredLeads(
  statusFilter?: ReviewStatus
): Promise<DiscoveredLead[]> {
  const params = new URLSearchParams();
  if (statusFilter) params.append("review_status", statusFilter);
  const url = params.toString()
    ? `/leads/discovered?${params}`
    : "/leads/discovered";
  const response = await apiClient.get<DiscoveredLead[]>(url);
  return response.data;
}

export async function reviewLead(
  leadId: string,
  action: ReviewStatus
): Promise<DiscoveredLead> {
  const response = await apiClient.post<DiscoveredLead>(
    `/leads/${leadId}/review`,
    { action }
  );
  return response.data;
}

export async function getScoreExplanation(
  leadId: string
): Promise<LeadScoreBreakdown> {
  const response = await apiClient.get<LeadScoreBreakdown>(
    `/leads/${leadId}/score-explanation`
  );
  return response.data;
}

export async function getPipeline(): Promise<PipelineSummary> {
  const response = await apiClient.get<PipelineSummary>("/leads/pipeline");
  return response.data;
}

export async function initiateOutreach(
  leadId: string,
  outreach: OutreachRequest
): Promise<OutreachResponse> {
  const response = await apiClient.post<OutreachResponse>(
    `/leads/outreach/${leadId}`,
    outreach
  );
  return response.data;
}
